### ISSUE TITLE
Java profiler setup to analyse performance

### ISSUE THREAD

**Issue by cmgrote** [odpi/egeria#814]
We should provide some instructions on configuring a Java profiler in order to delve into performance analysis detail of the various components (ie. to optimise areas like search, view OMAS, etc).

**Comment by planetf1** [comment]
Also enable this support in docker containers/helm chart for VDC through any necessary additional docs and/or container/chart changes

**Comment by ruxandraRosu** [comment]
[performance_profile.html.zip](https://github.com/odpi/egeria/files/2945164/performance_profile.html.zip)

**Comment by bogdan-sava** [comment]
interesting reports coming soon @ruxandraRosu and @popa-raluca

**Comment by ruxandraRosu** [comment]
[performance_profile_13_03.jps.zip](https://github.com/odpi/egeria/files/2961490/performance_profile_13_03.jps.zip)

**Comment by cmgrote** [comment]
@ruxandraRosu can you provide more details on what you were running and how you set it up?  Per @planetf1 's earlier comment, it would be great for us to provide a generic way of doing this so that everyone can do some profiling of their own code as part of individual testing.

**Comment by ruxandraRosu** [comment]
The tested flow is for the creation of a view triggered by a Semantic Assignment done to a column in IGC. 
This includes: 
- the retrieval of column context and publishing the event to virtualizer
- read event from virtualizer and create the view in Atlas
The setup included the following components:
-IGC with igc-repo-proxy
-Atlas with IGC metadata replicated to it
-OMASs server (with no local repository)
-virtualizer
-gaiandb
JPRofiler was installed and started on OMASs server host and attached to the process for OMASs server
from local JProfiler I've connected to remote Jprofiler process

**Comment by cmgrote** [comment]
> JPRofiler was installed and started on OMASs server host and attached to the process for OMASs server

Got it -- any chance we could explore an open source profiler as an alternative?  Not sure what's still around in the open source space (most I looked at seemed to be last updated 10+ years ago), but a couple still seem to be actively maintained, eg:

- https://visualvm.github.io/index.html
- https://eng.uber.com/jvm-profiler/

Basically the idea would be to document the instructions for taking an open source profiler and configuring Egeria to run with it (ie. in sufficient detail that Nigel and I could figure out how to build our Egeria docker container image to have an option of running with such a profiler turned on and available for remote connectivity and monitoring).

**Comment by planetf1** [comment]
++ to the above. I was going to say the same thing. As much as possibly it would be good if we could show how to use open source technology. It also means we can more easily discuss the results. Right now the Jprofiler output as is can't be interpreted without the tool installed - which requires license.

If there are particular observations to be discussed from the analysis you did perhaps we could track those in separate issue, as the insight may allow us to improve the code flow?

**Comment by ruxandraRosu** [comment]
The conclusions based on above report were the following:
1.main performance issue was caused by multiple RestTemplate instantiation/initialization. 
2.RestTemplate object should be created only one time and reused afterwards
3.using local repository for OMASs server(the new graph repository) could improve performance 
4.limit the use for java.lang.Class.forName
5.KafkaOpenMetadataTopicConnector could be configured to start either a consumer or a producer, not necessarily both
6.use log.isDebugEnabled() when log messages include formatting as this can be time consuming (the message is first formatted even if the level is INFO)
@mandy-chessell @bogdan-sava Please add any points I might have missed
I think we can run a new test considering 1->3 fixes were done to have a comparison. I can start looking into an open source technology as I ran into the same issue as you.

**Comment by planetf1** [comment]
It's worth noting that IntelliJ recently added a Java profiler. I'm not clear how usable it is remotely, nor can I compare to jprofiler from a 2 mins read, but it could be an option for those developers using intelij (probably most of us?). Though it may be ultimate edition only

https://blog.jetbrains.com/idea/2018/09/intellij-idea-2018-3-eap-git-submodules-jvm-profiler-macos-and-linux-and-more/

Oracle themselves offer the VisualVM profiler, though there may be license conditions (gpl)

**Comment by cmgrote** [comment]
Unfortunately the experimental profiler in IntelliJ does not seem to be in the community edition (so it's also not an open source option).

**Comment by planetf1** [comment]
Yes I realise that - but also know many (not all) of the current team have the tool already.
I checked with the adoptopenjdk team but it doesn't seem there's any open source suggestions from them, so this may be an area where proprietary is the main route taken

**Comment by ruxandraRosu** [comment]
The async-profiler seems to work with both Oracle jdk and openjdk, but I am not familiar with it yet, I will need to look into it.
https://github.com/jvm-profiling-tools/async-profiler
For Oracle jdk could we use Java Mission control with Flight Recorder? Or do we want to exclude it from the beginning as we're moving to openjdk?

**Comment by cmgrote** [comment]
The async-profiler looks promising -- even has a section on profiling within containers -- so I'd suggest pursuing that one first 😄

**Comment by planetf1** [comment]
It turns out that the profiler support in IntelliJ Ultimate is in fact async-profiler.
IntelliJ bundles the runtime, and then provides the GUI support around it -- that latter part is the ultimate added-value
Actually quite a nice approach as we can use Async profiler (if it's any good) and those that have ultimate can benefit from a better UI, but it's open to anyone.

I was able to run OMAGApplication under the profiler

Note that this worked with the openjdk hotspot build, but it does not work with OpenJ9

**Comment by planetf1** [comment]
For reference, the issue blocking use on OpenJ9 is https://github.com/eclipse/openj9/issues/5654

**Comment by cmgrote** [comment]
A few notes from my experimentation with async-profiler:

It does seem like the Flame Graphs may be the "way to go" to get some feel for where the major contributors to the overall runtime are...  Worth quickly scanning through the docs on how to read them (http://www.brendangregg.com/FlameGraphs/cpuflamegraphs.html), and in particular note:

- The width of each bar indicates its overall runtime contribution relative to the others -- so performance analysis is a matter of focusing in on the widest bars first.  (_Note in particular that there is no meaningful left-to-right ordering!_)
- This width is simply the overall runtime -- it does _not_ distinguish between a single run of the method that takes a long time and huge numbers of runs that each take relatively little time -- that's just the nature of sample-based profiling (how async-profiler works).
- Colour-coding is significant, but only to distinguish the runtime that's being measured, as it seems that the SVG output of async-profiler uses the Java profile: green is Java, yellow is C++, orange is kernel, and red is any remaining executing code (native user-level or other kernel modules)

This seems to imply there may be parts we can immediately zoom past in our analysis of our own code; but nonetheless may be interesting for broader context / impact of our code.  For example, this profile I generated from taking 30 seconds during the config and initialisation of the OMAG Server Platform with the IGC connector -- the reds and yellows to the right seem to be Java's internal workings on macOS (garbage collection, JIT compiling?, etc) while the greens to the left are the actual Java code we're eventually interested navigating (the specific Java thread that's running Apache Tomcat, within which other Apache components and Spring are running, and within which ultimately the OMAG Server Platform will be running):

<img width="1190" alt="Screenshot 2019-05-07 at 13 55 24" src="https://user-images.githubusercontent.com/5864283/57300991-db390880-70cf-11e9-8c40-177d1d614226.png">

While this means that there end up being a large amount of the call stack we can probably skip through before we get to what's "interesting" from our particular area, we can easily do that in the SVG by just clicking higher up in the call stack to "zoom in" on that area.

For example, here I've skipped up that left side of the call stack to narrow in on the point at which my calls to configure and activate the OMAG Server Platform with my configuration occur:

<img width="1187" alt="Screenshot 2019-05-07 at 14 01 13" src="https://user-images.githubusercontent.com/5864283/57301323-a8434480-70d0-11e9-9038-5c2811e2d8bd.png">

This process of "zooming in" then expands all sub-calls within that area to the full width of the SVG, so I immediately drop from focus the native JVM stuff that was on the very first layout.  And per the notes on Flame Graphs above I can see that the activation of the repository is what takes much longer than each of the configuration steps, and can start drilling into this `activateWithStoredConfig` or even a couple hops up the `initialize` call, to narrow my focus even further.

And if I want to highlight further where my code specifically sits in the overall scheme there's a little "Search" link in the upper-right you can click on to enter a simple regex for it to highlight.  In this example below, I've searched for "ibm.igc" and you can see all of those packages highlighted in purple:

<img width="1191" alt="Screenshot 2019-05-07 at 14 04 44" src="https://user-images.githubusercontent.com/5864283/57301542-33bcd580-70d1-11e9-969f-2341cbcf6b8e.png">

(As expected in my example, the largest contributors to the IGC connector startup time of several seconds appear to be a combination of REST calls and Java reflection.)

So for me it still looks useful -- particularly for free -- it just takes a little bit of navigation.  (Also this was when running everything directly on macOS: I tried with a Docker container, but there are some hoops to jump through there to setup the appropriate container security levels to actually be able to see all the same level of detail, particularly around other user-level, system-level and kernel-level calls...)

**Comment by planetf1** [comment]
Interesting summary - thanks.
A netflix prez on flame graphs and use for profiling -> https://www.usenix.org/conference/atc17/program/presentation/gregg-flame
Background at http://www.brendangregg.com/flamegraphs.html

Prez by those behind async-profiler - https://assets.ctfassets.net/oxjq45e8ilak/4mfbX5FJuw0A8M00UK4uKa/ce60f2cab12408e01ce927e90ebb2f7a/Andrey_Pangin__Vadim_Tsesko._The_Art_of_JVM_Profiling.pdf

**Comment by planetf1** [comment]
When we are profiling we also need to take care to do it in context. Anytime one looks at performance there are ALWAYS hotspots we can improve
 * What are the key use cases
 * Within the use cases what are the key interactions
 (for example time to service first request could be one - or we might decide it's not important)
 * Do we have SLAs/targets for RESPONSE TIMES
 * How about our expectations on memory
 * What are our scaling needs (here high cpu and memory can start to comprimise)
 * Can we use the tools to stop falling off a cliff (ie a trend that would lead to a jvm dump)

Agree though that this is a useful tool in the arsenal, and probably one that any developer could find useful to understand their own code, on top of purely a performance persuit.

**Comment by planetf1** [comment]
Also note that for the JVM functions - garbage collection etc, OpenJ9 is an alternative JVM - open source, available from adoptopenjdk.net that purports to significantly improve some aspects around memory allocation, startup. We shouldn't depend on it, but we may want to consider when we're looking more at performance.

**Comment by ruxandraRosu** [comment]
[report5 copy.html.txt](https://github.com/odpi/egeria/files/3196985/report5.copy.html.txt)

**Comment by planetf1** [comment]
For logging above it was mentioned to only use log.debug within a check that logging was enabled.

I agree that we should take care in particular in the normal flow (within error cases it's less important)

Since we're using slf4j an alternative is to redo the formatting using their parameter approach as per https://www.slf4j.org/faq.html#logging_performance - this would avoid the clutter if an additional if {}. Open for discussion .......

**Comment by cmgrote** [comment]
For a more in-depth and robust approach to profiling -- and one that can hopefully more seamlessly integrate into our existing development process (ie. plugin to IntelliJ) -- @jmertic and I are examining the potential to make use of the https://www.yourkit.com profiler under their open source license.

I will update further as we determine whether this is feasible or not; including (if so) documenting how best to leverage it as part of our development process.

**Comment by planetf1** [comment]
Good to see they seem to support a variety of jvms including those from adoptopenjdk with both hotspot and openj9 builds.

**Comment by jmertic** [comment]
Hi @cmgrote - apologizes as this fell through the cracks.

Yourkit has an open source license that requires the following...

```
To obtain a license a small free-form text endorsement should be
placed of the project web site.

The endorsement should contain:

1) Attached YourKit logo and text that your open source project is using YourKit
tools. Alternatively you can use use logo hosted at YourKit web-site
https://www.yourkit.com/images/yklogo.png

2) Acknowledgment should contain a hyperlink reference to YourKit web site.
```

**Comment by planetf1** [comment]
@cmgrote @jmertic Please let me know if we have any license for this now - would like to use

**Comment by jmertic** [comment]
Once #3353 is merged I can go back to them to get the license.

**Comment by planetf1** [comment]
it is now

**Comment by jmertic** [comment]
License at https://www.yourkit.com/m/manage?t=Y68A4AWYNXGUKAZQQ38VB9SDK99

**Comment by planetf1** [comment]
License now available - contact core team for info

**Comment by cmgrote** [comment]
Closing, as I think we have the tools we need now.
